### ISSUE TITLE
[Bug]: Discrepancies in the llama layer forward function between meta-llama, transformers and vLLM.

### ISSUE THREAD

**Issue by mcubuktepe** [vllm-project/vllm#12099]
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
Your output of `python collect_env.py` here
```

</details>

### 🐛 Describe the bug

Hello, I noticed that there are discrepancies in the forward pass of the attention layer between `llama`, `transformers`, and `vllm`

For example, in the `llama` [implementation](https://github.com/meta-llama/llama/blob/v2/llama/model.py#L406):

```
h = x + self.attention.forward(
    self.attention_norm(x), start_pos, freqs_cis, mask
)
out = h + self.feed_forward.forward(self.ffn_norm(h))
return out
```
RMSnorm is applied to the input, then attention is applied, before doing the residual step. Finally, the output is passed through a fully connected step with another RMSnorm normalization, and there's another residual operator before returning it.

In the `transformers` [implementation](https://github.com/huggingface/transformers/blob/v4.48.0/src/transformers/models/llama/modeling_llama.py#L327), we see a similar setup.

```
residual = hidden_states

hidden_states = self.input_layernorm(hidden_states)

# Self Attention
hidden_states, self_attn_weights, present_key_value = self.self_attn(
    hidden_states=hidden_states,
    attention_mask=attention_mask,
    position_ids=position_ids,
    past_key_value=past_key_value,
    output_attentions=output_attentions,
    use_cache=use_cache,
    cache_position=cache_position,
    position_embeddings=position_embeddings,
    **kwargs,
)
hidden_states = residual + hidden_states

# Fully Connected
residual = hidden_states
hidden_states = self.post_attention_layernorm(hidden_states)
hidden_states = self.mlp(hidden_states)
hidden_states = residual + hidden_states

outputs = (hidden_states,)

if output_attentions:
    outputs += (self_attn_weights,)

if use_cache:
    outputs += (present_key_value,)

return outputs
```

Specifically, the input is passed through an RMSnorm step, followed by attention and residual. After the residual, there's a fully connected step and a final residual before returning the output.

However, in vLLM, the residual is also passed through through the RMSnorm [here](https://github.com/huggingface/transformers/blob/v4.48.0/src/transformers/models/llama/modeling_llama.py#L327), i.e. this line:

```
hidden_states, residual = self.input_layernorm(
    hidden_states, residual)
```

Finally, the output is passed through the fully connected step, but the residual is not applied before returning the output:

```
# Fully Connected
hidden_states, residual = self.post_attention_layernorm(
    hidden_states, residual)
hidden_states = self.mlp(hidden_states)
return hidden_states, residual
```

This would be easy to fix, but I wanted to check whether this discrepancy is intentional by the vLLM team. Thanks a lot!

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
